### PR TITLE
Fix: after Audit changes

### PR DESCRIPTION
**Pull request by mrLSD**
## Description

Fixes for the reported issues. Includes:

➡️ **Bug:** In a `FixedPrice` sale with discount phases, a deposit that overshoots a phase/account/global limit by less than one purchasable sale token is fully refunded, yet integer rounding (`required_deposit` floors to `0`) still credits phantom weight to the depositor's allocation and to `total_sold_tokens`.

**Severity:** Not critical — no funds are lost or stolen (the deposit is returned in full); it only slightly over-counts sold tokens by sub-unit rounding, marginally reducing available sale capacity.

➡️ **Bug:** In a `FixedPrice` sale near the global cap, a deposit whose accepted weight converts to **zero** whole sale tokens (`calculate_amount_of_sale_tokens` floors to `0`) is partially charged instead of refunded, registering a paying participant (`participants_count`, `total_deposited`) with a zero-token allocation.

**Severity:** Not critical — no funds lost; a sub-unit "dust" deposit sticks and a zero-allocation participant is recorded. Fixed by treating a zero-sale-token result as a full refund.

➡️ **Bug:** The per-account minimum (`min_limit_per_account`) is checked against the **uncapped** discounted weight, so a phase/account/global cap can truncate the recorded position below the configured minimum after it has already "passed".

**Severity:** Not critical — bypass of a soft minimum-ticket policy; no funds lost, no oversell. Fixed by evaluating the minimum against the post-cap amount.

➡️ **Bug:** On a **partial** return from the intents account, `return_part_of_deposit` re-runs distribution on the unsent remainder and panics on `require!(refund == 0)`, reverting the callback and leaving the account stuck in `locked_withdraw` with the unsent amount dropped from accounting.

**Severity:** Not critical (latent — the standard intents contract never returns partially); if reached, a single-account lockout and loss of its unsent remainder. Fixed by re-crediting the leftover instead of panicking.

➡️ **Bug:** `finish_claim` treats an unparseable/oversized transfer result as a full refund (`read_ft_result` -> `None` -> `refund = assets_amount`), lowering `claimed` so the same allocation can be claimed again.

**Severity:** Not critical (latent — requires a non-standard sale token); if reached, a repeated claim. Fixed by failing closed — a delivered-but-unreadable transfer refunds nothing; only a failed transfer restores the claim.

➡️ **Bug:** `finish_mt_refund` panics on an empty result vector (`[]`), while the FT path returns the original amount — an asymmetric, fragile callback against a non-conformant intents contract.

**Severity:** Not critical (latent — a conformant NEP-245 intents contract never returns `[]`); robustness only. Fixed by mirroring the FT path (empty result ⇒ no charge).

## Summary by CodeRabbit

* **Bug Fixes**
  * Stabilized Price Discovery claim calculations during concurrent withdrawals.
  * Improved handling of failed, malformed, or ambiguous token transfer callbacks.
  * Prevented zero-token deposits from creating phantom investments or participants; deposits are fully refunded.
  * Corrected cap, minimum-limit, dust, and excess-deposit handling to prevent overselling.
  * Improved partial refund and withdrawal recovery behavior for FT and MT transfers.

* **Tests**
  * Added comprehensive regression coverage for claims, withdrawals, refunds, caps, and rounding edge cases.

### CODE REVIEW

File: contract/src/discount/state.rs [hunk 1]
```diff
@@ -237,6 +231,16 @@ impl DiscountState {
...
+            // Enforce the per-account minimum against the sale tokens this deposit is actually
+            // credited *after* caps, not the uncapped discounted amount: otherwise a deposit that
+            // clears the minimum before a phase/global/account cap truncates it would still be
+            // admitted below the minimum. Only an account's first qualifying deposit is checked;
+            // later top-ups stay exempt (see `is_min_limit_passed`).
+            let capped_sale_tokens = sale_tokens_per_deposit.saturating_sub(max_exceeded);
+            if !phase_params.is_min_limit_passed(capped_sale_tokens, existed_account_sale_tokens) {
+                continue;
+            }
+
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply contract/src/discount/state.rs:273]: @coderabbitai fixed 88568af6f3a2465ef74ad4a1cf699c388c3daeed

File: contract/src/mechanics/deposit.rs [hunk 2]
```diff
@@ -75,6 +75,14 @@ pub fn deposit(
         weight
     };
 
+    // A deposit that buys zero whole sale tokens (a sub-grain left by discount-phase or public-sale
+    // cap rounding at the configured price granularity) must not be charged: record nothing and
+    // refund it in full. Otherwise the buyer pays `deposit = amount - refund` for tokens they never
+    // receive, since `weight` already rounded down to zero sale tokens here.
+    if weight == 0 {
+        return Ok(amount);
+    }
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply contract/src/mechanics/deposit.rs:84]: @coderabbitai fixed 88568af6f3a2465ef74ad4a1cf699c388c3daeed

File: tests/src/tests/claim/intents.rs [hunk 3]
```diff
@@ -426,3 +435,111 @@ async fn claim_with_sale_tokens_refund() {
...
+            .unwrap()
+            .0;
+        if status == "Success" {
+            if sold < 10_000 {
+                let _ = bob.claim_to_intents(lp.id(), bob.id()).await;
+            }
+            break;
+        }
+        tokio::time::sleep(Duration::from_millis(40)).await;
+    }
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply tests/src/tests/claim/intents.rs]: @coderabbitai fixed 88568af6f3a2465ef74ad4a1cf699c388c3daeed

File: contract/src/lib.rs [hunk 4]
```diff
@@ -82,6 +82,14 @@ pub struct AuroraLaunchpadContract {
...
     /// The discounts state includes state for every discount phase.
     discount_state: Option<DiscountState>,
+    /// Frozen `total_sold_tokens`, captured on the first claim after the sale reached a success
+    /// status with no withdrawal in flight. Stable denominator for `PriceDiscovery` pro-rata claims,
+    /// immune to an in-flight withdrawal's transient decrement of `total_sold_tokens`.
+    final_total_sold: Option<u128>,
+    /// Number of withdrawals currently in flight (they decrement `total_sold_tokens` in
+    /// `do_withdraw` and the callback is still pending). `total_sold_tokens` is trustworthy as the
+    /// final claim denominator only when this is `0`.
+    withdraws_in_flight: u32,
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply contract/src/lib.rs]: @coderabbitai fixed 88568af6f3a2465ef74ad4a1cf699c388c3daeed

File: contract/src/mechanics/deposit.rs [hunk 5]
```diff
@@ -75,6 +75,14 @@ pub fn deposit(
         weight
     };
 
+    // A deposit that buys zero whole sale tokens (a sub-grain left by discount-phase or public-sale
+    // cap rounding at the configured price granularity) must not be charged: record nothing and
+    // refund it in full. Otherwise the buyer pays `deposit = amount - refund` for tokens they never
+    // receive, since `weight` already rounded down to zero sale tokens here.
+    if weight == 0 {
+        return Ok(amount);
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply contract/src/mechanics/deposit.rs:83]: @coderabbitai fixed 88568af6f3a2465ef74ad4a1cf699c388c3daeed

File: contract/src/claim.rs [hunk 6]
```diff
@@ -288,8 +311,16 @@ impl AuroraLaunchpadContract {
             "Expected one promise result"
         );
 
-        let refund =
-            read_ft_result(0).map_or(assets_amount, |used| assets_amount.saturating_sub(used));
+        // A failed transfer was not delivered, restore the full claim so it can be retried. A
+        // successful transfer whose result is oversized or unparseable (non-standard sale token) is
+        // treated as fully used (fail closed), so a delivered claim can never be replayed.
+        let refund = read_ft_result(0).map_or_else(
```
Comment (aleksuss) [inline contract/src/claim.rs]: I don't get the point of these changes, TBH. 

1. We already check the promise result in the `read_ft_result` function. We just convert a `Result::Err` to `Option::None`.
2. You pass 0 as the `max_len` argument to the `env::promise_result_checked`, which will return an error since it's a result of the `ft_transfer_call` transaction, which returns a `U128` -> (not zero length). 
 
But at the same time, the changes prove that we don't have an e2e test which covers this logic.

File: contract/src/claim.rs [hunk 7]
```diff
@@ -314,14 +337,16 @@ impl AuroraLaunchpadContract {
...
+            read_ft_result(0, assets_amount)
         } else {
             // A plain ft_transfer returns no value: a successful promise means nothing was
             // refunded, while a failed promise refunds the whole amount.
-            env::promise_result_checked(0, 0).map_or(assets_amount, |_| 0)
+            env::promise_result_checked(0, 0).map_or(0, |_| assets_amount)
         };
 
+        let refund = assets_amount.saturating_sub(consumed);
+
```
[the remark this answers is not in the document]
Comment (aleksuss) [reply contract/src/claim.rs:349]: A non-empty response from the `ft_transfer` could be possible only with a malicious token. A user has no possibility to modify the response of a `ft_transfer`, unlike `ft_transfer_call`.

**Review by aleksuss — APPROVED** [review]
